### PR TITLE
Add global beforeEach/afterEach hooks

### PR DESCRIPTION
**Pull request by brapifra**
## In a nutshell
This PR adds 4 new global parameters: `beforeEach`, `afterEach`, `asyncBeforeEach`and`asyncAfterEach` that will be executed during the execution of any property.
It solves the issue https://github.com/dubzzz/fast-check/issues/876

✔️ New feature
❌ Fix an issue
❌ Documentation improvement
❌ Other: *please explain*

(✔️: yes, ❌: no)

## Potential impacts

None

### CODE REVIEW

File: test/unit/check/property/AsyncProperty.spec.ts
Comment (dubzzz) [reply test/unit/check/property/AsyncProperty.spec.ts:12]: 🤔 Actually we may even mock `readConfigureGlobal` and don't use any of `configureGlobal` or `resetConfigureGlobal`. But you can keep it that way for this PR

File: test/unit/check/property/AsyncProperty.spec.ts
Comment (dubzzz) [inline test/unit/check/property/AsyncProperty.spec.ts:139]: Given (see other comments):

1. `beforeEach` and `asyncBeforeEach` should not be defined together in global settings (otherwise it will make the property fail to build)
2. For async properties the selected hook coming from global will be chosen as follow: take `asyncBeforeEach`, if not set take `beforeEach`
3. Users should be able to override them at anytime directly at call-site

My test suite would be different:
- Should take asyncBeforeEach coming from globals as the default for beforeEach if specified
- Should take beforeEach coming from globals as the default for beforeEach if specified
- Should throw if both beforeEach and asyncBeforeEach are defined in globals
- Should be able to override `beforeEach` globals locally

File: test/unit/check/property/Property.spec.ts
Comment (dubzzz) [inline test/unit/check/property/Property.spec.ts:12]: Same use `afterEach` instead

File: test/unit/check/property/Property.spec.ts
Comment (dubzzz) [inline test/unit/check/property/Property.spec.ts:127]: Given (see other comments):

1. `beforeEach` and `asyncBeforeEach` should not be defined together in global settings (otherwise it will make the property fail to build)
2. For async properties the selected hook coming from global will be chosen as follow: take `asyncBeforeEach`, if not set take `beforeEach`
3. Users should be able to override them at anytime directly at call-site

My test suite would be different:
- Should take beforeEach coming from globals as the default for beforeEach if specified
- Should throw if asyncBeforeEach is defined in globals
- Should throw if both beforeEach and asyncBeforeEach are defined in globals
- Should be able to override `beforeEach` globals locally

**Review by dubzzz — CHANGES_REQUESTED** [review]
Great job @brapifra!

I put some comments in order to ensure we preserve consistency across the primitives offered by the API.

File: test/unit/check/runner/configuration/GlobalParameters.spec.ts
Comment (brapifra) [reply test/unit/check/runner/configuration/GlobalParameters.spec.ts:19]: Ooops, yeah. I was using while developing and I forgot to remove it. I'll revert the changes.

File: test/unit/check/property/AsyncProperty.spec.ts
Comment (brapifra) [reply test/unit/check/property/AsyncProperty.spec.ts:12]: Yeah, we could mock it. Although I usually prefer to write integration-like tests. But it's really a personal preference so I don't mind changing it :smile: 
Regarding the `beforeEach`/`afterEach` comment. I'm pretty sure jest runs each test module/file in isolation. Therefore, I think it doesn't really matter to use one or the other in this case...

File: src/check/property/AsyncProperty.generic.ts
Comment (brapifra) [reply src/check/property/AsyncProperty.generic.ts:54]: Yeah, I initially implemented it like that. However, I soon realised that that kind of solution wouldn't fit my needs from the issue #876 
e.g. Let's say I have project with tons of property-based tests. Some of them are using beforeEach and/or afterEach. Now I want to add a call to a `cleanup` method globally via one of this new global parameters (e.g. afterEach). With your implementation, I would have to manually add that `cleanup` method to all these properties that were already implementing one/both hook(s).
Quite annoying IMO :sweat_smile:

File: src/check/runner/configuration/GlobalParameters.ts
Comment (brapifra) [reply src/check/runner/configuration/GlobalParameters.ts:15]: Not sure I understand 1 :thinking: 
What's the problem with being defined together?

File: src/check/property/AsyncProperty.generic.ts
Comment (dubzzz) [reply src/check/property/AsyncProperty.generic.ts:54]: Actually not really, my point is that if you define a value for `beforeEach` (the synchronous one) even asynchronous properties will take them into account.

File: src/check/runner/configuration/GlobalParameters.ts
Comment (dubzzz) [reply src/check/runner/configuration/GlobalParameters.ts:15]: In reality, the existence of both `property` (synchronous) and `asyncProperty` (asynchronous) is there for legacy reasons. Ideally all properties should be defined as asynchronous.

For instance, if you take a look into [this wrapper around fast-check designed for ava](https://github.com/dubzzz/ava-fast-check/), there is no more distinction between synchronous and asynchronous properties. Behind the scene everything is asynchronous as it only uses asyncProperty (but users don't know).

Comment (dubzzz) [reply src/check/runner/configuration/GlobalParameters.ts:15]: If you define the two global hooks together, there is probably an issue in your test suite. Or maybe hooks should not have been defined at a global level but at file or directory level.

My point is:
- If no global: property/asyncProperty take what the user specify using the beforeEach/afterEach methods
- If only global beforeEach: property/asyncProperty take it but replace it if the user specify one using the beforeEach method
- If only global asyncBeforeEach: property throws / asyncProperty takes it but replaces it if the user specify one using the beforeEach method
- If global beforeEach and asyncBeforeEach: property/asyncProperty throw

File: src/check/property/AsyncProperty.generic.ts
Comment (brapifra) [reply src/check/property/AsyncProperty.generic.ts:54]: Oh I see what you mean now. So in an example like this:
```typescript
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
```
It should print:
```
global beforeEach
global AfterEach
local afterEach
```
Right?

File: src/check/runner/configuration/GlobalParameters.ts
Comment (brapifra) [reply src/check/runner/configuration/GlobalParameters.ts:15]: As mentioned in https://github.com/dubzzz/fast-check/pull/900#discussion_r473277266, I think that it is valuable to not replace the global hooks when new ones have been defined with the beforeEach/afterEach methods.
If I have this case:
```typescript
configureGlobal({
    beforeEach: () => console.log('cleanup'),
});

fc.property(() => {}).beforeEach(() => console.log('beforeEach'));
```
I want both hooks to be run. Otherwise I'm going to have to call `cleanup` every time I use the `beforeEach` method in a property/asyncProperty.
Although I understand your point, an API like the one you suggest wouldn't solve my problem unfortunately :sweat_smile:

File: src/check/property/AsyncProperty.generic.ts
Comment (dubzzz) [reply src/check/property/AsyncProperty.generic.ts:54]: My idea would have been the following:

```js
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
```

Would print:
```
global beforeEach
local afterEach
```
_as global afterEach would have been overriden at call site level_

**AND I need to think about your idea:**

>  it is valuable to not replace the global hooks when new ones have been defined with the beforeEach/afterEach methods

Initially global overrides have been created as a syntaxic sugar to help users define only once a behaviour while it allows to override it locally. If we go for your proposal it means that in the case of beforeEach hooks, override will no be possible. I need to think abour it and come back to you (I will have a look to what Jest or Mocha do on this subject).

Meanwhile if we go in that direction I think that the ordering should be something like : 
- global before
- local before
- local after
- global after

Here are other examples:

<details>
<summary>Global beforeEach, Global afterEach</summary>

```js
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    afterEach: () => console.log('global afterEach')
});

fc.property(() => {});
// global beforeEach
// global afterEach

fc.asyncProperty(() => {});
// global beforeEach
// global afterEach

fc.property(() => {}).afterEach(() => console.log('local afterEach'));
// global beforeEach
// local afterEach

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
// global beforeEach
// local afterEach
```
</details>

<details>
<summary>Global beforeEach, Global asyncAfterEach</summary>

```js
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.property(() => {});
// OPT1 throws: Async hooks are not supported for synchronous properties. You may override it locally to prevent this error message
// OPT2 throws: Async hooks are not supported for synchronous properties. Please remove your global override or convert it to a synchronous hook or switch to an asynchronous property

fc.asyncProperty(() => {});
// global beforeEach
// global afterEach

fc.property(() => {}).afterEach(() => console.log('local afterEach'));
// OPT1
// global beforeEach
// local afterEach
// OPT2 throws: Async hooks are not supported for synchronous properties. Please remove your global override or convert it to a synchronous hook or switch to an asynchronous property

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
// global beforeEach
// local afterEach
```
</details>

<details>
<summary>Global afterEach, Global asyncAfterEach</summary>

```js
configureGlobal({
    afterEach: () => console.log('global afterEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.property(() => {});
// throws: Cannot define both sync and async version of the same hook

fc.asyncProperty(() => {});
// throws: Cannot define both sync and async version of the same hook
```
</details>

Comment (dubzzz) [reply src/check/property/AsyncProperty.generic.ts:54]: Concerning your point, what do you think of something like?

```js
fc.asyncProperty(() => {})
  .afterEach((previousAfterEach) => {
    previousAfterEach();
    console.log('local afterEach');
  });
```

or

```js
fc.asyncProperty(() => {})
  .afterEach(() => console.log('local afterEach'), { behaviour: 'keep-existing' });
// behaviour being one of 'keep-existing', 'override' (the default)...
```

or

```js
fc.asyncProperty(() => {})
  .afterEachKeepExisting(() => console.log('local afterEach'));
```

Comment (brapifra) [reply src/check/property/AsyncProperty.generic.ts:54]: Nice! Thanks for thinking about it :slightly_smiling_face: 
I personally find the API of option 1 a bit nicer. However, I think option 2 would give the flexibility of being able to modify the default behaviour via global config:
```typescript
fc.configureGlobal({ defaultBehaviourOfPropertyHooks: 'keep-existing' });
```
But maybe it's over-complicating the API, not sure... I'll leave it to you to decide! :smile:

Comment (dubzzz) [reply src/check/property/AsyncProperty.generic.ts:54]: Let's go for option 1! Actually it was my favourite when I wrote the comment and you seem to like it too:

```js
fc.asyncProperty(() => {})
  .afterEach((previousAfterEach) => {
    previousAfterEach();
    console.log('local afterEach');
  });
```

**Comment by brapifra** [comment]
I've made most of the requested changes. Let me know what you think @dubzzz !
The only thing that I didn't implement was the error throwing when a synchronous property is run and there is some async hook set up globally. Maybe I'm missing something but... why would we want to do that?
IMO synchronous property should simply ignore those async hooks instead of throwing an error. e.g.
```typescript
configureGlobal({
     asyncBeforeEach: () => {
          console.log("I just want to run this with my async properties and I don't want to break the sync ones");
     }
})
```

File: src/check/property/AsyncProperty.generic.ts
Comment (dubzzz) [inline src/check/property/AsyncProperty.generic.ts:65]: As users are now powerful enough to tell the framework how to merge multiple hooks together, I would make no differences between global and local hooks. And then let the user does what she/he wants with consecutive declarations of beforeEach. In other words, the global hook is just used to initialize beforeEachHook but it does not have any specific priviledges.

I believe a code like that:
```js
configureGlobal({ asyncBeforeEach: () => console.log('1') })

fc.asyncProperty(...)
  .beforeEach(prev => {
    prev();
    console.log('2');
  })
  .beforeEach(prev => {
    prev();
    console.log('3');
  })
  .beforeEach(prev => {
    console.log('4');
    prev();
    console.log('5');
  })
```

Should resulting into a hook that would do:
```js
() => {
    console.log('4');
    console.log('1');
    console.log('2');
    console.log('3');
    console.log('5');
}
```

In other words, I'd update the initialization part as follow:

```suggestion
  static dummyHook: AsyncPropertyHookFunction = (globalHook) => globalHook();
  private beforeEachHook: AsyncPropertyHookFunction;
  private afterEachHook: AsyncPropertyHookFunction;
  constructor(readonly arb: Arbitrary<Ts>, readonly predicate: (t: Ts) => Promise<boolean | void>) {
    const { asyncBeforeEach,  asyncAfterEach,  beforeEach,  afterEach  } = readConfigureGlobal() || {};
    if (asyncBeforeEach !== undefined && beforeEach !== undefined) {
      // Concerning those checks they need to be introduce as soon as we release the version.
      // As a library authors, it's easier to relax constraints that add new ones (breaking change)
      throw new Error('Illegal');
    }
    if (asyncAfterEach !== undefined && afterEach !== undefined) {
      throw new Error('Illegal');
    }
    this.beforeEachHook = asyncBeforeEach || beforeEach || AsyncProperty.defaultHook;
    this.afterEachHook = asyncAfterEach || afterEach || AsyncProperty.defaultHook;
  }
```

The beforeEach method would then have to be updated as follow:
```js
  beforeEach(hookFunction: AsyncPropertyHookFunction): AsyncProperty<Ts> {
    const previousBeforeEachHook = this.beforeEachHook;
    this.beforeEachHook = () => hookFunction(previousBeforeEachHook);
    return this;
  }
```

File: src/check/property/Property.generic.ts
Comment (dubzzz) [inline src/check/property/Property.generic.ts:14]:
```suggestion
export type PropertyHookFunction = (previousHookFunction: PropertyHookFunction) => void;
```

File: src/check/property/Property.generic.ts
Comment (dubzzz) [inline src/check/property/Property.generic.ts:72]: As for the async version:
```suggestion
    const { asyncBeforeEach,  asyncAfterEach,  beforeEach,  afterEach  } = readConfigureGlobal() || {};
    if (asyncBeforeEach !== undefined) {
      // Why this check?
      // - Users might have been told that a clean-up hook was globally set. So write her/his property
      //   thinking that the hook would be there. But the hook was actually defined on async versions only so it passes
      // - Consistency with the rest of the philosophy of the lib and where it goes concerning sync vs async
      // - As a library authors, it's easier to relax constraints that add new ones (breaking change). So putting it now
      // would at least require a change given there are good reasons not to do that. Not putting it would be impossible
      // to change
      throw new Error('Illegal');
    }
    if (asyncAfterEach !== undefined) {
      throw new Error('Illegal');
    }
    this.beforeEachHook = beforeEach || Property.defaultHook;
    this.afterEachHook = afterEach || Property.defaultHook;
```

File: src/check/property/AsyncProperty.generic.ts
Comment (dubzzz) [inline src/check/property/AsyncProperty.generic.ts:16]:
```suggestion
export type AsyncPropertyHookFunction =
  | ((previousHookFunction: AsyncPropertyHookFunction) => Promise<unknown>)
  | ((previousHookFunction: AsyncPropertyHookFunction) => void);
```

File: src/check/runner/configuration/GlobalParameters.ts
Comment (dubzzz) [inline src/check/runner/configuration/GlobalParameters.ts:21]: Directly merge Extra* in the GlobalPaarmeters as follow:

```suggestion
export type GlobalParameters = Pick<Parameters<unknown>, Exclude<keyof Parameters<unknown>, 'path' | 'examples'>> &
  {
  beforeEach?: GlobalPropertyHookFunction;
  afterEach?: GlobalPropertyHookFunction;
  asyncBeforeEach?: GlobalAsyncPropertyHookFunction;
  asyncAfterEach?: GlobalAsyncPropertyHookFunction;
};
```

It would be better in terms of generated doc.

File: src/check/runner/configuration/GlobalParameters.ts
Comment (dubzzz) [inline src/check/runner/configuration/GlobalParameters.ts:8]: As any exported type that are used by types or functions exposed to the end users they will need to be marked as public, documented and exported in fast-check-default.ts file.

**Review by dubzzz — COMMENTED** [review]
Once again thanks a lot for this huge contribution. I added some comments and suggestions to update the code and offer more flexibility and sanity checks to the end-users

File: src/check/property/AsyncProperty.generic.ts
Comment (brapifra) [reply src/check/property/AsyncProperty.generic.ts:16]: `previousHookFunction` can't be an `AsyncPropertyHookFunction`. This function doesn't accept any input arguments. That's why I used `GlobalAsyncPropertyHookFunction` instead.

File: src/check/property/Property.generic.ts
Comment (brapifra) [reply src/check/property/Property.generic.ts:72]: Fair enough! I'll change it

File: src/check/property/Property.generic.ts
Comment (brapifra) [reply src/check/property/Property.generic.ts:14]: Same here (https://github.com/dubzzz/fast-check/pull/900#discussion_r478601102)

File: src/check/property/AsyncProperty.generic.ts
Comment (brapifra) [reply src/check/property/AsyncProperty.generic.ts:65]: Oh, okay. I didn't know you wanted to go this route. Sounds good to me! :slightly_smiling_face:

File: src/check/property/AsyncProperty.generic.ts
Comment (brapifra) [inline src/check/property/AsyncProperty.generic.ts:58]: Happy to change these error messages if needed :slightly_smiling_face:

File: test/unit/check/property/AsyncProperty.spec.ts
Comment (dubzzz) [reply test/unit/check/property/AsyncProperty.spec.ts:12]: Just to make sure, prefer a `afterEach` here

File: src/check/runner/configuration/GlobalParameters.ts
Comment (dubzzz) [inline src/check/runner/configuration/GlobalParameters.ts:28]: Add a quick documentation to explain what are `beforeEach` and others?

Suggestion:
```suggestion
  /**
  Specify a function that will be called before each execution of a property.
  It behaves as-if you manually called `beforeEach` method on all the properties you execute with fast-check.
  
  The function will be used for both {@link fast-check#property} and {@link fast-check#asyncProperty}.
  This global override should never be used in conjunction with `asyncBeforeEach`.
  */
  beforeEach?: GlobalPropertyHookFunction;
  /**
  Specify a function that will be called after each execution of a property.
  It behaves as-if you manually called `afterEach` method on all the properties you execute with fast-check.
  
  The function will be used for both {@link fast-check#property} and {@link fast-check#asyncProperty}.
  This global override should never be used in conjunction with `asyncAfterEach`.
  */
  afterEach?: GlobalPropertyHookFunction;
  /**
  Specify a function that will be called before each execution of an asynchronous property.
  It behaves as-if you manually called `beforeEach` method on all the asynchronous properties you execute with fast-check.
  
  The function will be used only for {@link fast-check#asyncProperty}. It makes synchronous properties created by {@link fast-check#property} unable to run.
  This global override should never be used in conjunction with `beforeEach`.
  */
  asyncBeforeEach?: GlobalAsyncPropertyHookFunction;
  /**
  Specify a function that will be called after each execution of an asynchronous property.
  It behaves as-if you manually called `afterEach` method on all the asynchronous properties you execute with fast-check.
  
  The function will be used only for {@link fast-check#asyncProperty}. It makes synchronous properties created by {@link fast-check#property} unable to run.
  This global override should never be used in conjunction with `afterEach`.
  */
  asyncAfterEach?: GlobalAsyncPropertyHookFunction;
```

File: test/unit/check/property/Property.spec.ts
Comment (dubzzz) [reply test/unit/check/property/Property.spec.ts:12]:
```suggestion
  afterEach(() => resetConfigureGlobal());
```

I think it would be safer to always clean properly after the run. Cleaning before do not ensure we clean stuff for next tests.

**Review by dubzzz — COMMENTED** [review]
@brapifra Thank you so much for your work and time. Just a few remarks before we merge it. Really really thank for the time you already spent for this PR 👍

**Comment by dubzzz** [comment]
Perfect 👍 

Great job! Thanks a lot for the Pull Request!
I'll try to release a new minor including this new feature very soon.

**Pull request merged**
